drivers/crypto/crypto_it8xxx2_sha.c

```diff
@@ -0,0 +1,225 @@
+	0x650a7354, 0x766a0abb, 0x81c2c92e, 0x92722c85, 0xa2bfe8a1, 0xa81a664b,
+	0xc24b8b70, 0xc76c51a3, 0xd192e819, 0xd6990624, 0xf40e3585, 0x106aa070,
+	0x19a4c116, 0x1e376c08, 0x2748774c, 0x34b0bcb5, 0x391c0cb3, 0x4ed8aa4a,
+	0x5b9cca4f, 0x682e6ff3, 0x748f82ee, 0x78a5636f, 0x84c87814, 0x8cc70208,
+	0x90befffa, 0xa4506ceb, 0xbef9a3f7, 0xc67178f2
+};
+
+static void it8xxx2_sha256_init(bool init_k)
+{
+	int i;
+
+	chip_ctx.total_len = 0;
+	chip_ctx.w_input_index = 0;
+
+	/* Initialize hash values */
+	for (i = 0; i < ARRAY_SIZE(sha256_h0); i++) {
+		chip_ctx.h[i] = sha256_h0[i];
+	}
+	/* Initialize array of round constants */
+	if (init_k) {
+		for (int i = 0; i < ARRAY_SIZE(sha256_k); i++) {
+			chip_ctx.k[i] = sha256_k[i];
+		}
+	}
+}
+
+static void it8xxx2_sha256_module_calculation(void)
+{
+	uint32_t key;
+	uint8_t hash_ctrl;
+
+	/*
+	 * Since W field on it8xxx2 requires big-endian format, change byte
+	 * order before computing hash.
+	 */
+	for (int i = 0; i < SHA_SHA256_BLOCK_LEN_WORDS; i++) {
+		chip_ctx.w_sha[i] = sys_cpu_to_be32(chip_ctx.w_sha[i]);
+	}
+	/*
+	 * Global interrupt is disabled because the CPU cannot access memory
+	 * via the DLM (Data Local Memory) bus while HW module is computing
+	 * hash.
+	 */
+	key = irq_lock();
+	hash_ctrl = sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_HASHCTRLR);
+	sys_write8(hash_ctrl | IT8XXX2_SHA_START_SHA256,
+			IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_HASHCTRLR);
+	hash_ctrl = sys_read8(IT8XXX2_SHA_REGS_BASE + IT8XXX2_REG_HASHCTRLR);
+	irq_unlock(key);
+
+	chip_ctx.w_input_index = 0;
+}
+
+static int it8xxx2_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt,
+				bool finish)
+{
+	uint32_t rem_len = pkt->in_len;
+	uint32_t in_buf_idx = 0;
+
+	while (rem_len--) {
+		chip_ctx.w_input[chip_ctx.w_input_index++] =
+						pkt->in_buf[in_buf_idx++];
+		if (chip_ctx.w_input_index >= SHA_SHA256_BLOCK_LEN) {
+			it8xxx2_sha256_module_calculation();
+		}
+	}
+	chip_ctx.total_len += pkt->in_len;
+
+	if (finish) {
+		uint32_t *ob_ptr = (uint32_t *)pkt->out_buf;
+
+		/* Pre-processing (Padding) */
+		memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		chip_ctx.w_input[chip_ctx.w_input_index] = 0x80;
+
+		if (chip_ctx.w_input_index >= 56) {
+			it8xxx2_sha256_module_calculation();
+			memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		}
+		chip_ctx.w_sha[15] = sys_cpu_to_be32(chip_ctx.total_len * 8);
+		it8xxx2_sha256_module_calculation();
+
+		for (int i = 0; i < SHA_SHA256_HASH_LEN_WORDS; i++) {
+			ob_ptr[i] = sys_be32_to_cpu(chip_ctx.h[i]);
+		}
+
+		it8xxx2_sha256_init(false);
+	}
+
+	return 0;
+}
+
+static int it8xxx2_hash_session_free(const struct device *dev,
+				struct hash_ctx *ctx)
+{
+	it8xxx2_sha256_init(false);
+
+	return 0;
+}
+
+static inline int it8xxx2_query_hw_caps(const struct device *dev)
+{
+	return (CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
```

Comment:
Should this be a `#define`?

Reply:
Since it links to the query_hw_caps api, it should be a function.

Reply:
Oh I missed that, sorry.